### ISSUE TITLE
Analyzing the Gremlin attack endpoint

### ISSUE THREAD

**Issue by sonnysideup** [sonnysideup/go-gremlin#6]
I decided to take a step back from coding out the `CreateAttack()` func and really examine Gremlin's Attack API endpoint and collect my thoughts here.

Here are couple of sample request bodies from their docs:

```json
{
  "command": { "type": "blackhole", "args": [ "-i", "10.0.0.0/24" ] },
  "target": { "type": "Random" }
}

{
  "command": { "type": "blackhole", "args": [ "-i", "10.0.0.0/24" ] },
  "target": { "type": "Random", "tags": { "service": "api" } }
}

{
  "command": { "type": "blackhole", "args": [ "-i", "10.0.0.0/24" ] },
  "target": { "type": "Exact", "exact": ["client1", "client2"] },
  "labels": { "service": "api" }
}
```

Even though these JSON documents are well-structured, it's obvious that they are not *strongly-typed*. There is not a specific format for each of the different attack types and so it's left up to the user to figure which fields and flags are appropriate for the different attacks they want to launch. That being said, I think there's 3 approaches we can take here.

1. Figure out what needs to be done in order to build a strong type system around the loose data definition that the API expects. This most likely will require an explicity implementation of the [Marshaler Interface](https://golang.org/pkg/encoding/json/#Marshaler) since we'll need to coerce a non-standard representation of the struct in JSON.
2. Pass in a raw JSON string as the body. This gets the job done but is super ugly and error prone.
3. Find a middle ground and make it work for now, then improve it with time.

Can you sense which option I'm leaning towards right now? 😉 

A good first cut, albeit overly verbose and not safe in the sense that it allows invalid user input, is to represent the command structure explicitly, like so:

```go
// Attack command details
type Command struct {
	// Type should be one of the following: blackhole, cpu, io, latency, memory,
	// package_loss, shutdown, dns, time_travel, disk, process_killer
	Type string `json:"type"`

	// Args supplied to the command should be identical to those passed to the CLI
	// and UI. Please note that some commands have no required args.
	Args []string `json:"args,omitempty"`
}

// Attack target details
type Target struct {
	// Type should either be "Random" or "Exact"
	Type string `json:"type"`

	// Exact list of hosts to target
	Exact []string `json:"exact,omitempty"`

	// Tags restrict an attack only to hosts with the corresponding kv tags.
	Tags map[string]string `json:"tags,omitempty"`
}

// Encapsulates the details required to launch an attack
type AttackCommand struct {
	Command `json:"command"`
	Target  `json:"target"`

	// Labels are used to target Docker containers running on target hosts
	Labels map[string]string `json:"labels,omitempty"`
}
```

Someone can then create an attack command using those structs.

```go
        // represents the last example up above
	at := &AttackCommand{
		Command{
			Type: "blackhole",
			Args: []string{"-i", "10.0.0.0/24"},
		},
		Target{
			Type:  "exact",
			Exact: []string{"client1", "client2"},
		},
		map[string]string{"service": "api"},
	}
```

In summary, I really don't like this approach but it will work well enough to unblock @ljanke and give us more time to design a type-safe interface and figure out if/how to cleanly coerce our structs into the JSON format we need via a custom marshal implementation or some other mechanism.

/cc @sebradloff

**Comment by ljanke** [comment]
I think this could be the low level interface to a "GenericAttack" method that we could build a somewhat typed API for each attack type on top of it. This will be fairly easy for the attacks with just a few parameters but not easy or clean for attacks with many parameters. The main complication seems to be that Go doesn't have good support for optional or named parameters. I'm don't think there is going to be a clean way to get a completely typed API (i.e. checkable by the compiler at compile time) for the more complicated attacks given that. 

For attacks with many potential parameters some of which are optional, the problem seems similar to what Cobra faces when it's dealing with a CLI that takes many possible options. So maybe we need something similar here: A command registers it's argument values, whether they are optional, their type, and then they are checked, and serialized at runtime.

Regarding the proposal for AttackCommand: I think it can be a good interim starting point to build a more typed API on, but I think the Command and Target should probably be interfaces.

/cc @sebradloff

**Comment by sebradloff** [comment]
@ljanke I agree with the want to create a GenericAttack and then inherit that for creating specific attack types. @sonnysideup and I spent a decent amount of time figuring out how to do that in Golang, but couldn't figure it out. I' new to Golang so there might be an inheritance pattern out there.

What if we designed the attacks similarly to NewClient, with the idea being that we could implement some sort of logic on initialization that would confirm that the attack type specified can take the args given? Just spit balling, I haven't given it a ton of thought atm.

To start, I'm good with the AttackCommand proposal and I'm not sure what the pros and cons are for implementing an interface over struct at the Command and Target level.

**Comment by sonnysideup** [comment]
The case may be that we have to use one or several of these approaches:
- custom marshalling
- custom tag validation (really cool article: https://sosedoff.com/2016/07/16/golang-struct-tags.html)
- some form of composition
- some form of interface magic

I think this is just going to require a bit of research and prototyping to get right.

**Comment by ljanke** [comment]
Or we could just switch it all to Python? 🤨
